### PR TITLE
Introduce weak vendors mode

### PR DESCRIPTION
**Pull request by greg0ire**
Deprecations coming from the vendors are segregated from other
deprecations. A new mode is introduced, in which deprecations coming
from the vendors are not taken into account when deciding to exit with
an error code.

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/7453

Sample output : 

# Weak vendor mode

## With both vendors and non-vendor errors

```

WARNINGS!
Tests: 1068, Assertions: 2714, Warnings: 6, Skipped: 15.

Remaining deprecation notices (1)

some error: 1x
    1x in SonataAdminBundleTest::testBuild from Sonata\AdminBundle\Tests

Remaining vendor deprecation notices (4)

Legacy deprecation notices (48)
```

Exit code is 1

## After fixing non-vendor errors

```
WARNINGS!
Tests: 1068, Assertions: 2714, Warnings: 6, Skipped: 15.

Remaining vendor deprecation notices (4)

Legacy deprecation notices (48)
```

Exit code is 0

# TODO

- [x] fix colorization issues (vendor deprecation notices are always in red)
- [x] make the vendor detection more robust
- [x] make the vendor dir configurable
- [x] ~figure out how to run tests and~ add more of them
    - [x] test on non-vendor file
    - [x] test on vendor file
- [x] do not change the output of other modes

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I see two issues here:
- the first, minor, is that the vendor idr is not always called "vendors"
- the second is more serious: any code in vendors is always called by some code in userland. Code don't get called magically - but because userland asked for it - thus is not such thing as "vendor deprecations" vs others. Or do you have a way to decide that work reliably?

**Comment by greg0ire** [comment]
1. Yes that's an issue (see my todo list), maybe I can work around that with another env var?
2. Not an issue IMO, I'm only looking at the first file of the stack trace, which is the one where `trigger_error` is written.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: @nicolas-grekas : we're outside the while on `$trace` here, so really, we're not looking at every file of the stack trace, but only at the file that has the error triggered in it.

**Comment by greg0ire** [comment]
Third issue, even more minor, some people might have a `vendor` directory inside their `src` (why not). Might be solvable with `__DIR__` though.

**Comment by nicolas-grekas** [comment]
> I'm only looking at the first file of the stack trace

That means you target deprecations triggered by code in src/. Ok.

For vendors, see my PR about ComposerResource.

**Comment by greg0ire** [comment]
> I'm only looking at the first file of the stack trace

Actually that was wrong and has nothing to do with the stack trace, I'm looking at the `$errfile` argument of the error handler function. So if I have a file in `src`, that calls a function from the `vendor` dir, that has a `trigger_error` in it, `$errfile` will be `vendor/callee.php`, and not `src/caller.php`, right?

**Comment by greg0ire** [comment]
> For vendors, see my PR about ComposerResource.

[link for the lazy](https://github.com/symfony/symfony/pull/21505)

**Comment by ogizanagi** [comment]
[link for the laziest](https://github.com/nicolas-grekas/symfony/blob/e930f4064460bcfa4b1c19fc37881354dac920a2/src/Symfony/Component/Config/Resource/ComposerResource.php#L85-L93) 😄

**Comment by greg0ire** [comment]
Maybe I should explain a bit more the goal of this PR, I feel I'm being unclear.

This is going to used way more in libs than in project, that seldom use deprecation notices themselves. @nicolas-grekas I think that's why you talk about userland vs vendors.

Let there be lib X, that has n dependencies. Anytime any of the n dependencies introduces a new deprecation, I have to fix it, and *no PR can be merged during that time*, unless I use `SYMFONY_DEPRECATIONS_HELPER=weak`.

And that's what I do most of the time, because I don't want to shift the burden of fixing these to anyone who wants to contribute anything unrelated to the lib, and has bad luck doing so at the wrong time.

Now let's suppose a contributor submits a new PR, and that PR introduces a deprecation notice in *the codebase of the lib* , and by that I don't mean a deprecated call to another lib, I mean a call to `trigger_error` directly. Then they must be careful to mark all corresponding tests with `@group legacy`, and fix all calls throughout the codebase of the lib so that they use the new way of doing things.

In my experience, many people forget to do that, and deprecations are introduced, that could easily be avoided. See [this PR](https://github.com/sonata-project/SonataAdminBundle/pull/4310/files) where I end up doing the work that should have been done by the contributor from the start.

**Comment by nicolas-grekas** [comment]
Doc PR would be really nice before merging since the use case should be explained there :)

**Comment by stof** [comment]
@greg0ire the issue is that your logic looks wrong. Vendor deprecations must not be distinguished based on the location of the deprecation, but based on the caller location.
You will almost never call your own deprecated code, as you are also the one deprecating it. However, you are very likely to be calling a deprecated method of one of your dependency. In this case, ``trigger_error`` is inside the vendor folder, but the place needing to be fixed is inside your own code.
And using the backtrace to identify whether it is called from your own code or no is fragile, because your code may not be the direct caller. We have cases in Symfony where we have more classes between your own code and the code triggering the deprecation for instance, while the fix is still needed in your own code, for instance when asking to migrate from a type name to a class name to reference form types in Symfony 2.8

**Comment by greg0ire** [comment]
> You will almost never call your own deprecated code, as you are also the one deprecating it

That's precisely the case I'm trying to help fix however. There are contributors that contribute new deprecations, and forget to fix these calls. Then it get merged, but could have easily be prevented.

**Comment by greg0ire** [comment]
> Doc PR would be really nice before merging since the use case should be explained there :)

Didn't notice your comment. Will do!

**Comment by greg0ire** [comment]
[Here is the docs PR](https://github.com/symfony/symfony-docs/pull/7453)

**Comment by nicolas-grekas** [comment]
Thanks, just posted a few short notes

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (rybakit) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Missing `true` as a 3rd parameter for `in_array`: https://3v4l.org/hsgBm

**Review by rybakit — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: fixed, thanks!

**Comment by greg0ire** [comment]
I changed the code so that other modes output is no longer split into vendors / non-vendors.

**Comment by greg0ire** [comment]
I fixed the colorization issue, deprecation notices are always in yellow now.

**Comment by fabpot** [comment]
fabbot reports errors that should be fixed.

**Comment by greg0ire** [comment]
@fab{b,p}ot  : fixed

**Comment by greg0ire** [comment]
@ogizanagi : should I do something similar to [this](https://github.com/symfony/symfony/pull/21505/files#diff-a10c9afd8feaccd1ff66ffd54e2835bcR1500)? Also, not sure what `ComposerResource::getVendors()` returns… does it return one entry per library? Or can there be several `vendor` dirs (named differently of course)?

**Comment by ogizanagi** [comment]
> not sure what ComposerResource::getVendors() returns… does it return one entry per library? Or can there be several vendor dirs (named differently of course)?

Yes, one entry per vendor directory, as you can have multiple vendor dirs. For instance the global composer install one + the project one. Personally, I'm always in this situation because I use `auto_prepend_file=${HOME}/.composer/vendor/autoload.php` in my php.ini file.

> I don't understand what the strpbrk(substr($path, strlen($vendor), 1), '/'.DIRECTORY_SEPARATOR) part is about.

I think it's about checking the given path is **in** the vendor dir, and not just the dir itself. `strpbrk` is probably used rather than `in_array` or anything else for some performances reasons. But actually, I'm not 100% sure of the intents. @nicolas-grekas would answer you far better than me 😄

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Here, I choose to put the assignment inline, to avoid this : 

```php
if (DeprecationErrorHandler::MODE_WEAK_VENDORS === $mode) {
    // let's compute the variable only when it is needed.
    $isVendor = self::inVendors($file);
}
// rest of the code
```

**Comment by greg0ire** [comment]
Can someone explain to me how to run the tests for this particular class? I've found `./src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/*.phpt` but can't figure out how these are used.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: you can't use this class: it's not a dep of the bridge (and we don't want to add it as a dep to the bridge)
use copy/paste instead :)

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: the in_array + line splits look unreadable to me - what about using two comparisons on a single line instead?

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: should be on a single line - please remove the bracket arount the assignation, but add some around the logical operators

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: `remaining from vendors`?

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: double comparison - in_array provides no value here to me

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Damn, you're right! Will do.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: I think it's a matter of taste, because to me the more readable version is of course mine. Plus it's less repetitive. I'll do as you say though, b/c you're probably more aware of what the sf style is.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: You guys don't like linebreaks, do you? :P Will do.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: would give the output "remaining from vendors deprecations". I think "remaining vendor deprecations" is better, right?

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: The value is: not repeating `$group`, and being ready for more groups. The downside is the `true` argument… will change as you say but again, matter of taste here IMO.

**Comment by greg0ire** [comment]
How can I tell fabbot not to try to fix purposefully broken code? 

```diff
diff -ru src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_non_vendor.phpt src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_non_vendor.phpt
--- src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_non_vendor.phpt	2017-02-14 19:47:58.381296941 +0000
+++ src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_non_vendor.phpt	2017-02-14 19:47:58.985277835 +0000
@@ -31,14 +31,14 @@
     public function testLegacyFoo()
     {
         @trigger_error('silenced foo deprecation', E_USER_DEPRECATED);
-        trigger_error('unsilenced foo deprecation', E_USER_DEPRECATED);
-        trigger_error('unsilenced foo deprecation', E_USER_DEPRECATED);
+        @trigger_error('unsilenced foo deprecation', E_USER_DEPRECATED);
+        @trigger_error('unsilenced foo deprecation', E_USER_DEPRECATED);
     }
 
     public function testNonLegacyBar()
     {
         @trigger_error('silenced bar deprecation', E_USER_DEPRECATED);
-        trigger_error('unsilenced bar deprecation', E_USER_DEPRECATED);
+        @trigger_error('unsilenced bar deprecation', E_USER_DEPRECATED);
     }
 }
 
diff -ru src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_vendor.phpt src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_vendor.phpt
--- src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_vendor.phpt	2017-02-14 19:47:58.525292385 +0000
+++ src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_vendor.phpt	2017-02-14 19:47:59.193271256 +0000
@@ -24,13 +24,13 @@
     public function testLegacyFoo()
     {
         @trigger_error('silenced foo deprecation', E_USER_DEPRECATED);
-        trigger_error('unsilenced foo deprecation', E_USER_DEPRECATED);
+        @trigger_error('unsilenced foo deprecation', E_USER_DEPRECATED);
     }
 
     public function testNonLegacyBar()
     {
         @trigger_error('silenced bar deprecation', E_USER_DEPRECATED);
-        trigger_error('unsilenced bar deprecation', E_USER_DEPRECATED);
+        @trigger_error('unsilenced bar deprecation', E_USER_DEPRECATED);
     }
 }
```

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Can anyone think of less invasive way of testing this? It's a bit invasive…

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: I hope you have nothing about the multiline `$colorize` call. IMO it's more readable in an objective way (you can see all the code w/o scrolling when reviewing, for instance).

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: When testing, `$path` equals `-`, and I had no idea how to change that, so I added a public property dedicated to tests, because [testing trumps encapsulation](https://8thlight.com/blog/uncle-bob/2015/06/30/the-little-singleton.html), says Uncle Bob.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (stof) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: We should rather write a test creating a fake vendor folder with its autoloading file (similar to the one generated by Composer).

Having this argument here means that your test will **not** ensure that the feature is working (try to replace the normal logic with just ``return false`` and see you tests passing)

**Comment by greg0ire** [comment]
An issue I also have locally : https://travis-ci.org/symfony/symfony/jobs/201634200#L1851

> Fatal error: Uncaught Error: Class 'PHPUnit_Util_Test' not found in /home/travis/build/symfony/symfony/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php:114

No idea why this class can't be loaded…

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: > We should rather write a test creating a fake vendor folder with its autoloading file (similar to the one generated by Composer).

Ok, I'll try to do something like that.

**Comment by greg0ire** [comment]
> How can I tell fabbot not to try to fix purposefully broken code?

I can see that there are no required checks, so I guess you can just ignore it when merging?

**Comment by greg0ire** [comment]
Pushed a new, less intrusive test based on @stof . Now I just need to figure out why the low deps build fails.

**Comment by greg0ire** [comment]
I don't reproduce the failing build with

```
composer update --prefer-lowest
cp  src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php vendor/symfony/phpunit-bridge/
phpunit src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak_vendors_on_vendor.phpt
```

Why not?

**Comment by greg0ire** [comment]
I finally reproduced that locally on a travis docker image.

**Comment by greg0ire** [comment]
Looks like the error handler is the `getMode()` closure when this happens.

**Comment by greg0ire** [comment]
Ok so I got more details with `error_get_last()`

```
+array(4) {
+  ["type"]=>
+  int(1)
+  ["message"]=>
+  string(50) "Cannot access self:: when no class scope is active"
+  ["file"]=>
+  string(85) "/home/travis/builds/me/symfony/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php"
+  ["line"]=>
+  int(92)
+}
```

**Comment by greg0ire** [comment]
This is ready for review.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Having this public is a bit crappy, I think I'll try using a closure instead.

**Comment by greg0ire** [comment]
I pushed a new version with `inVendors` as a closure, so that it is not exposed publicly.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: borrowed from ContainerBuilder:
`if (0 === strpos($path, $vendor) && false !== strpbrk(substr($path, strlen($vendor), 1), '/'.DIRECTORY_SEPARATOR)) {`

**Review by nicolas-grekas — COMMENTED** [review]
LGTM, one remaining comment + a rebase to see tests green and we're all good I think (see also fabbot's failures)

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: What is the second part about exactly? I was told it was about checking whether `$path` was a vendor dir itself, which wouldn't be useful in that case, would it?

**Comment by greg0ire** [comment]
@nicolas-grekas rebase done, fixable fabbot.io errors fixed, and code snippet copied.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: it's to check that the prefix ends in with a directory separator

**Comment by greg0ire** [comment]
The low deps build is unhappy. There must be something non php 5.3 compliant in there…

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Oh so we check that `$path` isn't `/path/to/vendorThisIsQuiteMalicious.php`?

**Comment by greg0ire** [comment]
@nicolas-grekas I got rid from more references to `self` that can't work with php 5.3, using the same trick as in the other closure. There's also a new commit that `var_dump`s `error_get_last()` in case of deprecation error failure. You can see how it looks like here : https://travis-ci.org/symfony/symfony/jobs/202907646#L4552

**Comment by greg0ire** [comment]
I rebased again, can anyone review and remove the "Needs work" label? This is supposed to be finished.

**Comment by greg0ire** [comment]
Changed the poor man's error handler from `var_dump` to `print_r` after discussing with @Nek-

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: unrelated, and very fragile to me, should be removed from this PR

Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Ok let's remove it.

Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Not sure what exactly qualifies as fragile here.

Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: the link between `error_get_last()` and the error handler change - maybe I'm wrong - it's still unrelated so please open another PR if you want to investigate the idea

Comment (greg0ire) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Unrelated indeed. Will do

**Comment by greg0ire** [comment]
This is supposed to be ready on my end, please review again.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Review by xabbuh — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @greg0ire.

**Comment by greg0ire** [comment]
Made a separate PR for the poor man's error handling : #21795

**Pull request merged**
